ocaml/gencert/pem.ml

```diff
@@ -60,24 +62,33 @@ let key =
   key_footer kind *> return (String.concat "" [header kind; body; footer kind])
   <?> "key"
 
-let line = take_till is_eol *> end_of_line
-
-(* try to read a key, or skip a line and try again *)
-let until_key = fix (fun m -> key <|> line *> m) <?> "until_key"
-
 let cert =
   cert_header >>= fun hd ->
   data >>= fun body ->
   cert_footer >>= fun tl -> return (String.concat "" [hd; body; tl]) <?> "cert"
 
-(* try to read a cert, or skip a line and try again *)
-let until_cert = fix (fun m -> cert <|> line *> m) <?> "until_cert"
+let line = take_till is_eol *> end_of_line
+
+let any_block =
+  cert >>= (fun c -> return (Cert c)) <|> (key >>= fun k -> return (Key k))
+
+(* this skips over junk until we succeed finding the next block *)
+let block = fix (fun m -> any_block <|> line *> m) <?> "until_block"
+
+(* collect and tag all blocks *)
+let blocks = many block <?> "PEM blocks"
 
+(* decompose blocks into certs and keys *)
 let pem =
-  until_key >>= fun private_key ->
-  until_cert >>= fun host_cert ->
-  many until_cert >>= fun other_certs ->
-  many end_of_line *> return {private_key; host_cert; other_certs} <?> "pem"
+  let strip = function Cert c -> c | Key k -> k in
+  blocks >>= fun bs ->
+  match List.partition (function Key _ -> true | Cert _ -> false) bs with
+  | [Key k], Cert c :: xs ->
```

Comment:
The code assumes that the first certificate is the one signed by the key. This might not be true.

I think for now this is fine, but it might not always be true